File: TTesting.R
library(genefilter)
library(GGally)
library(ggplot2)

source("MaclearnUtilities.R")

## source("LoadData.R")
## source("NormalizeData.R")
## source("RestrictData.R")
## source("ExtractYs.R")
load("prepared_datasets.RData")

ynums = lapply(ys, function(y) {
    ynames = names(y)
    y = as.numeric(y) - 1
    names(y) = ynames
    return(y)
})


## -----------------------------------------------------------------
## t.test example (using equal variance t test)
## -----------------------------------------------------------------
shengene = xnorms$shen[ , "NM_008161"]
shengene_nervous = shengene[ys$shen == "TRUE"]
shengene_other = shengene[ys$shen == "FALSE"]
t.test(shengene_nervous, shengene_other, var.equal=TRUE)

cor.test(shengene, ynums$shen, method="pearson")


## -----------------------------------------------------------------
## t tests for all genes in shen set
## -----------------------------------------------------------------
tShenAll = colttests(as.matrix(xnorms$shen), ys$shen)
tShenAll$q.value = p.adjust(tShenAll$p.value, method="fdr")
## let's try something else...
xscShen = scale(xnorms$shen, center=TRUE, scale=TRUE)
summary(colMeans(xscShen))
summary(colSds(xscShen))
yscShen = scale(ynums$shen)
tShenAll$pearson = as.numeric( (t(yscShen) %*% xscShen) / (length(yscShen)-1) )
## sort by p.value
tShenAll = tShenAll[order(tShenAll$p.value), ]

plot(tShenAll$pearson, tShenAll$p.value, log='y', pch=16, cex=0.5)


## -----------------------------------------------------------------
## t tests for all genes in each set
## -----------------------------------------------------------------
tTestResults = mapply(
    FUN = function(x, y) {
        out = colttests(as.matrix(x), y)
        out$q.value = p.adjust(out$p.value, method="fdr")
        out$pearson = as.numeric(
            (t(scale(as.numeric(y))) %*% as.matrix(scale(x))) /
            (length(y)-1)
        )
        out = out[order(out$p.value), ]
        return(out)
    },
    xnorms,
    ys,
    SIMPLIFY = FALSE
)


## -----------------------------------------------------------------
## let's look at top genes in each set
## -----------------------------------------------------------------
lapply(tTestResults, head)


boxstrip(
    xnorms$shen[ rownames(tTestResults$shen)[1:9] ],
    ys$shen,
    colscale = c("black", "red")
)

boxstrip(
    xnorms$patel[ rownames(tTestResults$patel)[1:9] ],
    ys$patel,
    colscale = c("black", "red")
)

boxstrip(
    xnorms$montastier[ rownames(tTestResults$montastier)[1:9] ],
    ys$montastier,
    colscale = c("black", "red")
)

boxstrip(
    xnorms$hess[ rownames(tTestResults$hess)[1:9] ],
    ys$hess,
    colscale = c("black", "red")
)


## -----------------------------------------------------------------
## generate fancy p.value vs pearson correlation plot
## -----------------------------------------------------------------
ggdata = do.call(
    rbind,
    args = lapply(X=names(tTestResults), FUN=function(setname) {
        tres = tTestResults[[setname]]
        data.frame(
            gene = rownames(tres),
            set = paste0(setname, " (", nrow(xnorms[[setname]]), ")"),
            tres
        )
    })
)
ggdata$`|t|` = abs(ggdata$statistic)
ggdata$set = factor(as.character(ggdata$set),
                    levels=levels(ggdata$set)[order(sapply(xnorms, nrow))])

ggobj = ggplot(data=ggdata,
        mapping=aes(x=pearson, y=`|t|`, color=set))
ggobj = ggobj + ylim(c(0, 10))
ggobj = ggobj + geom_line()
ggobj = ggobj + scale_color_manual(
        values=c("darkred", "red", "darkgray", "black"))
ggobj = ggobj + theme_bw()
## pdf("TStatVsPearson.pdf", h=5, w=5*1.45)
print(ggobj)
## garbage = dev.off()


## -----------------------------------------------------------------
## generate fancy p.value vs pearson correlation plot-
## -----------------------------------------------------------------
ggobj = ggplot(data=ggdata,
        mapping=aes(x=pearson, y=p.value, color=set))
ggobj = ggobj + scale_y_log10()
ggobj = ggobj + geom_line()
ggobj = ggobj + scale_color_manual(
        values=c("darkred", "red", "darkgray", "black"))
ggobj = ggobj + theme_bw()
## pdf("PValVsPearson.pdf", h=5, w=5*1.45)
print(ggobj)
## garbage = dev.off()



## -----------------------------------------------------------------
## complementary features
## -----------------------------------------------------------------
compResults = gramSchmidtSelect(x=xnorms$patel, y=ys$patel, g="NAMPT")
compFeats = names(compResults[
        order(abs(compResults), decreasing=TRUE)])[1:1000]
compR2 = sapply(compFeats, function(g) {
    summary(lm(ynums$patel ~ xnorms$patel$NAMPT +
            xnorms$patel[[g]]))$r.squared
})
plot(compR2, type="l")

## model being optimized is truly linear model -- not
## a good choice for classification!
summary(lm(ynums$patel ~ xnorms$patel$NAMPT + xnorms$patel$SEC61G))
## but features likely to work well for glm classification
## (and other classification methods) as well
summary(glm(ys$patel ~ xnorms$patel$NAMPT + xnorms$patel$SEC61G, family=binomial))

File: TTesting.py
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import scipy as sp
import scipy.stats as stats
import sklearn as sk
import sklearn.preprocessing as preprocessing

plt.ion()
plt.style.use('fivethirtyeight')

import MaclearnUtilities
from MaclearnUtilities import bhfdr, colttests, gramSchmidtSelect

import RestrictedData
xs = RestrictedData.xs
xnorms = RestrictedData.xnorms
annots = RestrictedData.annots
ys = RestrictedData.ys
ynums = RestrictedData.ynums


## -----------------------------------------------------------------
## ttest example (using equal variance t test)
## -----------------------------------------------------------------
shengene = xnorms['shen']['NM_008161']
shengene_nervous = shengene[ys['shen']]
shengene_other = shengene[~ys['shen']]
tout = stats.ttest_ind(shengene_nervous.values,
                       shengene_other.values,
                       equal_var = True)

stats.pearsonr(shengene, ynums['shen'])


## -----------------------------------------------------------------
## t tests for all genes in shen set
## -----------------------------------------------------------------
tShenAll = colttests(xnorms['shen'], ynums['shen'])
tShenAll['q'] = bhfdr(tShenAll.p)
## let's try something else...
xscShen = preprocessing.scale(xnorms['shen'])
xscShen = pd.DataFrame(xscShen,
                      index = xnorms['shen'].index,
                      columns = xnorms['shen'].columns)
xscShen.mean(axis=0)
xscShen.std(axis=0)

yscShen = preprocessing.scale(ynums['shen'].astype('float'))
tShenAll['pearson'] = np.dot(yscShen, xscShen) / len(yscShen)
## sort by p
tShenAll.sort_values('p', inplace=True)


## -----------------------------------------------------------------
## t tests for all genes in each set
## -----------------------------------------------------------------
def tTestPlus(x, y):
    out = colttests(x, y)
    out['q'] = bhfdr(out['p'])
    out['pearson'] = np.dot(
        preprocessing.scale(y.astype('float')),
        preprocessing.scale(x)
    ) / len(y)
    out.sort_values('p', inplace=True)
    return out

tTestResults = {k : tTestPlus(xnorms[k], ynums[k]) for k in xnorms}


## -----------------------------------------------------------------
## generate fancy p vs pearson plot
## -----------------------------------------------------------------
plt.clf()
ax = plt.subplot(111)
colors = {
    "montastier" : "black",
    "patel" : "darkgray",
    "hess" : "red",
    "shen" : "darkred"
}
for s in tTestResults:
    plotdata = pd.DataFrame({'gene' : tTestResults[s].index,
                             'set' : s + " (" + str(xnorms[s].shape[0]) + ")",
                             'p' : tTestResults[s].p,
                             'pearson' : tTestResults[s].pearson})
    plotdata.sort_values("pearson", inplace=True)
    plotdata.plot(x = "pearson",
                  y = "p",
                  color = colors[s],
                  logy = True,
                  label = s + " (" + str(xnorms[s].shape[0]) + ")",
                  ax = ax)


compResults = gramSchmidtSelect(x = xnorms['patel'],
                                y = ynums['patel'],
                                g = 'NAMPT')
compResults = compResults.loc[
        compResults.abs().sort_values(ascending=False).index]

